### PR TITLE
Feature/authentication middleware

### PR DESCRIPTION
**Pull request by edwardlin1695**
Major Changes:

1. Introduced authentication middleware
2. Provides example adjustments to routing where middleware is integrated
3. Refactors claim extraction for tokens

### CODE REVIEW

File: api/route/route.go [hunk 1]
```diff
@@ -1,17 +1,20 @@
...
-	apiRouterGroup := engine.Group("/api")
-	NewLoginRouter(env, db, apiRouterGroup)
-	NewSignupRouter(db, apiRouterGroup)
+	publicApiRouterGroup := engine.Group("/api")
+	NewLoginRouter(env, db, publicApiRouterGroup)
+	NewSignupRouter(db, publicApiRouterGroup)
 	//template router
-	TemplateRouter(db, apiRouterGroup)
-	NewWorkspaceRouter(db, apiRouterGroup)
+	privateApiRouterGroup := engine.Group("/api")
```
Comment (StevenDB995) [inline api/route/route.go:16]: not very sure... but isn't it the same instance as publicApiRouterGroup?

File: model/jwt_custom.go [hunk 2]
```diff
@@ -3,11 +3,17 @@ package model
 import "github.com/golang-jwt/jwt/v4"
 
 type JwtCustomClaims struct {
-	Email string `json:"email"`
+	Email       string `json:"email"`
+	UID         uint   `json:"uid"`
```
Comment (StevenDB995) [inline model/jwt_custom.go:7]: Maybe UserID is a better naming :)

**Review by StevenDB995 — APPROVED** [review]
Looks good to me in general :)

File: api/route/route.go (hunk 1 above)
Comment (edwardlin1695) [reply api/route/route.go:16]: While they use the same path, they are grouped differently.
Only the privateApiRouterGroup will employ the middleware.

Comment (edwardlin1695) [reply api/route/route.go:16]: We can always create additional separation by adjusting the route for private and public routes

**Pull request merged**